builtin/common/serialize.lua

```diff
@@ -1,205 +1,190 @@
-	-- or not.  For instance, all of these chunks create nest points in
-	-- table x: "x = {}; x[x] = 1", "x = {}; x[1] = x",
-	-- "x = {}; x[1] = {y = {x}}".
-	-- To handle those, two tables are used by mark_nest_point:
-	-- * nested - Transient set of tables being currently traversed.
-	--   Used for detecting nested tables.
-	-- * nest_points - parent->{key=value, ...} table cantaining the nested
-	--   keys and values in the parent.  They're all dumped after all the
-	--   other table operations have been performed.
-	--
-	-- mark_nest_point(p, k, v) fills nest_points with information required
-	-- to remember that key/value (k, v) creates a nest point  in table
-	-- parent. It also marks "parent" and the nested item(s) as occuring
-	-- multiple times, since several references to it will be required in
-	-- order to patch the nest points.
-	local nest_points  = {}
-	local nested = {}
-	local function mark_nest_point(parent, k, v)
-		local nk, nv = nested[k], nested[v]
-		local np = nest_points[parent]
-		if not np then
-			np = {}
-			nest_points[parent] = np
-		end
-		np[k] = v
-		seen[parent] = 2
-		if nk then seen[k] = 2 end
-		if nv then seen[v] = 2 end
-	end
-
-	-- First phase, list the tables and functions which appear more than
-	-- once in x.
-	local function mark_multiple_occurences(x)
-		local tp = type(x)
-		if tp ~= "table" and tp ~= "function" then
-			-- No identity (comparison is done by value, not by instance)
-			return
-		end
-		if seen[x] == 1 then
-			seen[x] = 2
-		elseif seen[x] ~= 2 then
-			seen[x] = 1
-		end
+local table_insert, table_concat, string_dump, string_format, string_match, math_huge
+    = table.insert, table.concat, string.dump, string.format, string.match, math.huge
 
-		if tp == "table" then
-			nested[x] = true
-			for k, v in pairs(x) do
-				if nested[k] or nested[v] then
-					mark_nest_point(x, k, v)
-				else
-					mark_multiple_occurences(k)
-					mark_multiple_occurences(v)
-				end
-			end
-			nested[x] = nil
-		end
+-- Recursively counts occurences of objects (non-primitives including strings) in a table.
+local function count_objects(value)
+    local counts = {}
+	if value == nil then
+		-- Early return for nil
+		return counts
 	end
+    local function count_values(val)
+		local type_ = type(val)
+		if type_ == "boolean" or type_ == "number" then return end
+        local count = counts[val]
+        counts[val] = (count or 0) + 1
+        if not count and type_ == "table" then
+            for k, v in pairs(val) do
+                count_values(k)
+                count_values(v)
+            end
+        end
+    end
+    count_values(value)
+    return counts
+end
 
-	local dumped     = {}  -- object->varname set
-	local local_defs = {}  -- Dumped local definitions as source code lines
-
-	-- Mutually recursive local functions:
-	local dump_val, dump_or_ref_val
+-- Build a table with the succeeding character from A-Z
+local succ = {}
```

Comment:
Not intuitive variable name. Proposal: `next_reference_character`
Needs a comment what this is used for, and not what it does. Letter is already obvious.

Reply:
> Not intuitive variable name. Proposal: `next_reference_character`
> Needs a comment what this is used for, and not what it does. Letter is already obvious.

I disagree. See Haskell:

```
$ ghci
GHCi, version 8.6.5: http://www.haskell.org/ghc/  :? for help
Prelude> succ 'A'
'B'
```

Reply:
`succ` doesn't seem intuitive to me either. Especially for people who don't know Haskell, I think SmallJoker has a good suggestion.

Reply:
`succ_ref_char` it is.